spec/app/ics-021-nft-transfer/README.md

```diff
@@ -0,0 +1,360 @@
+---
+
+## Synopsis
+
+This standard document specifies packet data structure, state machine handling logic, and encoding details for the transfer of non-fungible tokens over an IBC channel between two modules on separate chains. The state machine logic presented allows for safe multi-chain denomination handling with permissionless channel opening. This logic constitutes a "non-fungible token transfer bridge module", interfacing between the IBC routing module and an existing asset tracking module on the host state machine.
+
+### Motivation
+
+Users of a set of chains connected over the IBC protocol might wish to utilise an asset issued on one chain on another chain, perhaps to make use of additional features such as exchange or privacy protection, while retaining non-fungibility with the original asset on the issuing chain. This application-layer standard describes a protocol for transferring non-fungible tokens between chains connected with IBC which preserves asset non-fungibility, preserves asset ownership, limits the impact of Byzantine faults, and requires no additional permissioning.
+
+### Definitions
+
+The IBC handler interface & IBC routing module interface are as defined in [ICS 25](../../core/ics-025-handler-interface) and [ICS 26](../../core/ics-026-routing-module), respectively.
+
+### Desired Properties
+
+- Preservation of uniqueness (two-way peg).
+- Preservation of total supply (maintained on a single source chain & module).
+- Permissionless token transfers, no need to whitelist connections, modules, or denominations.
+- Symmetric (all chains implement the same logic, no in-protocol differentiation of hubs & zones).
+- Fault containment: prevents Byzantine-inflation of tokens originating on chain `A`, as a result of chain `B`'s Byzantine behaviour (though any users who sent tokens to chain `B` may be at risk).
+
+## Technical Specification
+
+### Data Structures
+
+Only one packet data type is required: `NonFungibleTokenPacketData`, which specifies the denomination, amount, sending account, and receiving account.
+
+```typescript
+interface NonFungibleTokenPacketData {
+  classId: string
+  classUri: string
+  tokenId: string
+  tokenUri: string
+  sender: string
+  receiver: string
+}
+```
+
+As tokens are sent across chains using the ICS 21 protocol, they begin to accrue a record of channels for which they have been transferred across. This information is encoded into the `classId` field. 
+
+The ics21 token classes are represented in the form `{ics21Port}/{ics21Channel}/{classId}`, where `ics21Port` and `ics21Channel` are an ics21 port and channel on the current chain for which the token exists. The prefixed port and channel pair indicate which channel the token was previously sent through. If `{classId}` contains `/`, then it must also be in the ics21 form which indicates that this token has a multi-hop record. Note that this requires that the `/` (slash character) is prohibited in non-IBC token denomination names.
```

Comment:
Let's take an example:
+ On chain A we have 2 NFTs (n1, n2) with same class `k`.
+ Chain A has two channels to chain B: r1 and r2.
+ If we sent `n1` from A to B using r1 and `n2` using r2, then in chain B they will have different class id

Correct? This smells bad, as we increase fragmentation.

Reply:
The same thing happens with fungible tokens of the same denom; that is the way it is with IBC, and the fragmentation will be handled by UI providers.

Reply:
I know. But maybe there is a way to fix it / make it better.

Reply:
I think this design is pretty fundamental to how IBC works. Probably not the time to focus on revising such a core component when the fastest path would be to re-use the norm. That problem should likely be focused on holistically and then applied to both ICS-20 and ICS-721 at a future date. Would be curious to hear whether it's been discussed by the IBC team before (cc @crodriguezvega )

Reply:
OK, wanted to check if there are already solutions for that.

Reply:
This is core to how IBC works. Agreed to keep as is